### PR TITLE
cli: introduce --provider

### PR DESCRIPTION
**Pull request by cjp256**
Valid options include `host`, `lxd`, and `multipass`.  This approach
is more extendable than the --use-<provider> flags for future providers.

- Split _BUILD_OPTIONS into _BUILD_OPTIONS and _PROVIDER_OPTIONS
- Split add_build_options into add_build_options and add_provider_options
- Convert `clean` command to use `add_provider_options` so it doesn't
  duplicate provider flags (--use-lxd, --destructive-mode, etc.).

- Minor cleanup for BuilderEnvironmentConfig:
-- drop is_multipass, no-one uses it
-- document lxd as valid provider
-- rename force_provider to build_provider

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/cli/_options.py [hunk 1]
```diff
@@ -65,49 +65,97 @@ def prompt_for_value(self, ctx):
...
     ),
     dict(
         param_decls="--use-lxd",
         is_flag=True,
-        help="Forces snapcraft to use LXD to build.",
+        help="Forces snapcraft to use LXD to build (implies `--provider=lxd`).",
+    ),
+    dict(
+        param_decls="--provider",
+        help="Specify build provider - options include `lxd`, `host`, or `multipass`.",
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: better use the Choice type for this

File: snapcraft/cli/_options.py [hunk 2]
```diff
@@ -65,49 +65,97 @@ def prompt_for_value(self, ctx):
...
             "--destructive-mode",
             "--use-lxd and --destructive-mode cannot be used together.",
         )
-    elif force_use_lxd:
-        provider = "lxd"
-    elif force_destructive_mode:
-        provider = "host"
+
+    if provider:
+        if provider not in ["lxd", "host", "multipass"]:
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: by using the Choice type, this would not be needed

File: snapcraft/cli/_options.py [hunk 3]
```diff
@@ -65,49 +65,97 @@ def prompt_for_value(self, ctx):
...
+                "--provider",
+                "--destructive-mode cannot be used with `--provider={}`".format(
+                    provider
+                ),
+            )
+
+        if provider == "host" and not destructive_mode:
+            raise click.BadOptionUsage(
+                "--provider",
+                "`--provider=host` requires `--destructive-mode` to acknowledge side effects.",
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: good!

File: snapcraft/cli/_options.py [hunk 4]
```diff
@@ -65,49 +65,97 @@ def prompt_for_value(self, ctx):
...
-        provider = None
-    return env.BuilderEnvironmentConfig(force_provider=provider)
+        if use_lxd:
+            provider = "lxd"
+
+        if destructive_mode:
+            provider = "host"
+
+    # BuilderEnvironmentConfig will determine provider if None.
+    return env.BuilderEnvironmentConfig(build_provider=provider)
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: getting rid of BuildEnvironmentConfig might also be fun :-)

Comment (cjp256) [reply snapcraft/cli/_options.py]: Done! And with that also managed to consolidate SNAPCRAFT_BUILD_ENVIRONMENT handling in with the --provider option. Much nicer :D

File: snapcraft/cli/_options.py (hunk 2 above)
Comment (cjp256) [reply snapcraft/cli/_options.py]: Thanks for that "DUHH" moment.  Done!

File: snapcraft/cli/_options.py (hunk 1 above)
Comment (cjp256) [reply snapcraft/cli/_options.py]: Thanks for that "DUHH" moment. Done!

**Comment by cjp256** [comment]
Still need to add some tests...

**Comment by cjp256** [comment]
Still a TODO in the env for the tests... don't merge :D

File: snapcraft/cli/_options.py [hunk 5]
```diff
@@ -65,49 +65,112 @@ def prompt_for_value(self, ctx):
...
+                "--use-lxd cannot be used with --provider={}".format(provider)
+            )
+
+        if provider != "host" and destructive_mode:
+            raise click.BadArgumentUsage(
+                "--destructive-mode cannot be used with --provider={}".format(provider)
+            )
+
+        # Specifying --provider=host requires the use of --destructive-mode.
+        # SNAPCRAFT_BUILD_ENVIRONMENT=host does not.
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: maybe it should, our tests would need to adapt and we would need an MP for launchpad to use --destructive-mode

File: snapcraft/cli/_options.py [hunk 6]
```diff
@@ -65,49 +65,112 @@ def prompt_for_value(self, ctx):
...
     ),
     dict(
         param_decls="--use-lxd",
         is_flag=True,
-        help="Forces snapcraft to use LXD to build.",
+        help="Forces snapcraft to use LXD to build (implies `--provider=lxd`).",
+    ),
+    dict(
+        param_decls="--provider",
+        envvar="SNAPCRAFT_BUILD_ENVIRONMENT",
```
Comment (sergiusens) [inline snapcraft/cli/_options.py]: If https://github.com/snapcore/snapcraft/pull/2678/files#diff-1b5470a9f2e177fc1606479628098324R123 is done anyways, it probably shouldn't be here... I'd rather keep this envvar as internal, we should probably log.warning when being used (not this PR), but not expose it here either

Comment (cjp256) [reply snapcraft/cli/_options.py]: https://click.palletsprojects.com/en/7.x/api/
`show_envvar – controls if an environment variable should be shown on the help page. Normally, environment variables are not shown.`

**Comment by cjp256** [comment]
Rebased

File: tests/spread/build-providers/lxd-basic/task.yaml [hunk 7]
```diff
@@ -2,6 +2,7 @@ summary: Build a basic snap using LXD and ensure that it runs
 
 environment:
   SNAP_DIR: ../snaps/make-hello
+  SNAPCRAFT_BUILD_ENVIRONMENT: lxd
```
Comment (sergiusens) [inline tests/spread/build-providers/lxd-basic/task.yaml]: why do we need this? --use-lxd should be enough

File: snapcraft/cli/_options.py (hunk 6 above)
Comment (sergiusens) [reply snapcraft/cli/_options.py]: let's use this key

File: tests/spread/build-providers/lxd-basic/task.yaml (hunk 7 above)
Comment (cjp256) [reply tests/spread/build-providers/lxd-basic/task.yaml]: IIRC (and by that, I mean if the commit message is accurate lol)

SNAPCRAFT_BUILD_ENVIRONMENT: "host" is set by spread.yaml.  Which is now an error due to mismatch.  The option would be to either unset SNAPCRAFT_BUILD_ENVIRONMENT or setting it to "lxd".

Comment (sergiusens) [reply tests/spread/build-providers/lxd-basic/task.yaml]: Can we unset it then? It might bring the expectation that this is the way to do things

Comment (cjp256) [reply tests/spread/build-providers/lxd-basic/task.yaml]: Sure! Looks like that matches what was done for multipass too.

**Comment by cjp256** [comment]
Rebased.

**Pull request merged**